src/utils/makeReactNativeConfig.js

```diff
@@ -237,6 +248,62 @@ function makeReactNativeConfig(
   return config;
 }
 
+/**
+ * Creates an array of configs based on changing `env` for every
+ * platform and returns
+ */
+function makeReactNativeConfig(
+  userWebpackConfig: WebpackConfigFactory,
+  options: ConfigOptions,
+  platform: Platform
+) {
+  /**
+   * We should support also the old format of config
+   * 
+   * module.exports = {
+   *   entry: './index.js',
+   * };
+   * 
+   * module.exports = ({ platform }) => ({
+   *   entry: `./index.${platform}.js`,
+   * });
+   */
+  const isLegacy =
+    typeof userWebpackConfig === 'function' || !userWebpackConfig.webpack;
+
+  if (isLegacy) {
+    logger.warn(
+      'You using deprecated style of the configuration. Please follow the docs to the upgrade.'
+    );
+
+    return DEPRECATEDMakeReactNativeConfig(
+      userWebpackConfig,
+      options,
+      platform
+    );
+  }
+
+  const env = { ...options, platform };
+
+  const {
+    webpack: webpackConfigFactory /* , ...haulConfig */,
+  } = userWebpackConfig;
+
+  if (
+    typeof webpackConfigFactory !== 'function' &&
+    typeof webpackConfigFactory !== 'object'
+  ) {
+    throw new Error('You have provided a wrong Webpack configuration.');
```

Comment:
Make the error message something like "The webpack configuration must be an object or a function returning an object" and maybe add a link to the configuration docs.